internal/provisioner/helm/controllers/bundle_controller.go

```diff
@@ -0,0 +1,241 @@
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.9.2/pkg/reconcile
+func (r *BundleReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	l := log.FromContext(ctx)
+	l.V(1).Info("starting reconciliation")
+	defer l.V(1).Info("ending reconciliation")
+	bundle := &rukpakv1alpha1.Bundle{}
+	if err := r.Get(ctx, req.NamespacedName, bundle); err != nil {
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	u := updater.NewBundleUpdater(r.Client)
+	defer func() {
+		if err := u.Apply(ctx, bundle); err != nil {
+			l.Error(err, "failed to update status")
+		}
+	}()
+	u.UpdateStatus(updater.EnsureObservedGeneration(bundle.Generation))
+
+	finalizerResult, err := r.Finalizers.Finalize(ctx, bundle)
+	if err != nil {
+		u.UpdateStatus(
+			updater.EnsureResolvedSource(nil),
+			updater.EnsureContentURL(""),
+			updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+			updater.EnsureCondition(metav1.Condition{
+				Type:    rukpakv1alpha1.TypeUnpacked,
+				Status:  metav1.ConditionUnknown,
+				Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+				Message: err.Error(),
+			}),
+		)
+		return ctrl.Result{}, err
+	}
+	var (
+		finalizerUpdateErrs []error
+	)
+	// Update the status subresource before updating the main object. This is
+	// necessary because, in many cases, the main object update will remove the
+	// finalizer, which will cause the core Kubernetes deletion logic to
+	// complete. Therefore, we need to make the status update prior to the main
+	// object update to ensure that the status update can be processed before
+	// a potential deletion.
+	if finalizerResult.StatusUpdated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Status().Update(ctx, bundle))
+	}
+	if finalizerResult.Updated {
+		finalizerUpdateErrs = append(finalizerUpdateErrs, r.Update(ctx, bundle))
+	}
+	if finalizerResult.Updated || finalizerResult.StatusUpdated || !bundle.GetDeletionTimestamp().IsZero() {
+		err := apimacherrors.NewAggregate(finalizerUpdateErrs)
+		if err != nil {
+			u.UpdateStatus(
+				updater.EnsureResolvedSource(nil),
+				updater.EnsureContentURL(""),
+				updater.SetPhase(rukpakv1alpha1.PhaseFailing),
+				updater.EnsureCondition(metav1.Condition{
+					Type:    rukpakv1alpha1.TypeUnpacked,
+					Status:  metav1.ConditionUnknown,
+					Reason:  rukpakv1alpha1.ReasonProcessingFinalizerFailed,
+					Message: err.Error(),
+				}),
+			)
+		}
+		return ctrl.Result{}, err
+	}
+
+	unpackResult, err := r.Unpacker.Unpack(ctx, bundle)
+	if err != nil {
+		return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("source bundle content: %v", err))
+	}
+	if unpackResult.State != source.StateUnpacked {
+		return ctrl.Result{}, updateStatusUnpackFailing(&u, fmt.Errorf("unknown unpack state %q: %v", unpackResult.State, err))
+	}
+
+	// Writing chart contents in temp directory for linting
```

Comment:
Ohhh. Sorry I didn't see this earlier, but how about we do something along these lines:
```go
pr, pw := io.Pipe()
eg, ctx := errgroup.WithContext(ctx)
eg.Go(func() error {
	return pw.CloseWithError(util.FSToTarGZ(pw, bundleFS))
})
eg.Go(func() error {
	chart, err := loader.LoadArchive(pr)
	if err != nil {
		return err
	}
	return chart.Validate()			
})
return eg.Wait()
```

Reply:
Do we also need to run the linter? If there are lint errors, should we reject that bundle?

Reply:
It rejects the bundle if it fails lint now.  The validate function just validates the chart.yaml file.  Is it enough check before it stores the chart?

Reply:
I changed to just call `Validate`.